### ISSUE TITLE
Regression: assimp fails to parse well-formed XML

### ISSUE THREAD

**Issue by rhaschke** [assimp/assimp#4213]
**Describe the bug**
assimp fails to read well-formed COLLADA XML with self-closing XML tags:
```xml
    <contributor>
      <author/>
      <authoring_tool>FBX COLLADA exporter</authoring_tool>
    </contributor>
```
**To Reproduce**:
[meshes.zip](https://github.com/ros-visualization/rviz/files/7588080/meshes.zip)
```
$ assimp info /tmp/meshes/cube.dae 
Launching asset import ...           OK
Validating postprocessing flags ...  OK
ERROR: Failed to load file: Collada: /tmp/meshes/cube.dae - Expected end of `<author>` element.
assimp info: Unable to load input file /tmp/meshes/cube.dae
```

**Platform (please complete the following information):**
 - OS: Ubuntu 20.04
 - Version: `libassimp-dev:amd64 5.0.1~ds0-1build1`

**Additional context**
Works in `4.1.0~dfsg-3`

**Comment by umlaeute** [comment]
afaict, this was fixed with assimp-5.1.

running on my Debian/sid system:

~~~sh
$ assimp  version
------------------------------------------------------ 
Open Asset Import Library ("Assimp", https://github.com/assimp/assimp) 
 -- Commandline toolchain --
------------------------------------------------------ 

Version 5.1 -debug -shared -st (GIT commit 0)

$ for f in /tmp/meshes/*.*; do assimp info "${f}" >/dev/null; echo "$? $f"; done
0 /tmp/meshes/backpack.dae
0 /tmp/meshes/cube.dae
0 /tmp/meshes/drill.dae
0 /tmp/meshes/extinguisher.dae
0 /tmp/meshes/helmet.dae
0 /tmp/meshes/phone.dae
0 /tmp/meshes/rope.dae
0 /tmp/meshes/survivor.dae
0 /tmp/meshes/vent.dae

~~~

**Comment by rhaschke** [comment]
This is good news. When will 5.1 be released to Ubuntu 20.04?

**Comment by umlaeute** [comment]
i (being the Debian maintainer of `assimp`) don't intent to.

Ubuntu-20.04 was released 1½ years ago. typically the software it includes does not get updated afterwards (apart from security fixes).
Ubuntu-22.04 will most likely include `assimp-5.1` though.


this of course doesn't mean that you cannot use assimp-5.1 on Ubuntu-20.04 (just that there are no Ubuntu-official packages)

**Comment by rhaschke** [comment]
That's not so good news :disappointed: I never understood, why obvious bug fixes and regressions must not be fixed in a released Ubuntu distro. As long as the fix is API and ABI compatible, there isn't any risk in doing so, right?
Of course, ensuring ABI compatibility is not trivial, but there are tools to check it...

**Comment by kimkulling** [comment]
What you can do is building assimp on your own. It shall be really simple: https://github.com/assimp/assimp/blob/master/Build.md#manual-build-instructions

**Comment by rhaschke** [comment]
We are using assimp as a dependency in several software packages that are released as debians as well.
So, while I can manually install assimp, I can't enforce users of our software packages to do so.

**Comment by umlaeute** [comment]
> That's not so good news disappointed I never understood, why obvious bug fixes and regressions must not be fixed in a released Ubuntu distro. 

regardless, these are the rules for a distribution that offers stable releases.
so while we can complain about them, it doesn't change them.
esp. the assimp bugtracker is not the place to discuss the policy of Ubuntu.

> As long as the fix is API and ABI compatible, there isn't any risk in doing so, right?

no, in general this is unfortunately wrong.

software is a complex beast, and changing one line might have unforeseeable side-effects.

in the concrete case: what exactly is the change required to fix your problem?
so far we know that assimp-5.1 fixes the problem, so the change happened somewhere between 5.0.1 (which is shipped in Focal) and 5.1.1.
this gives us about 3000 commits:

~~~sh
$ git rev-list v5.0.1..v5.1.1 --count
3027
~~~

many of these commits depend on other commits (so it's not like we could pick e.g. 9e8dbd0ca549457db3b04ed4a27a09790e59122f and be done).
so should we just import assimp-5.1 wholesale?
afaict, assimp-5.1 has not broken ABI nor API, so according to your logic an update should be safe. but...

while it fixes the parsing of well-formed XML, it also introduces regressions that crash on virtual any unsupported file (#4201), crash on some DAE files (again #4201) and some LWO files (#4209); and the python bindings are broken (#4208). (these are just a few problems that *I* have found when running the Debian testsuite; there are probably more issues lurking, as i haven't even started doing any functional tests).

**Comment by rhaschke** [comment]
I agree that releasing a new version might introduce new issues. Usually, important bug fixes are thus cherry-picked to the original release branch, such that side effects as you describe them can be limited (or avoided). Of course, finding the exact commits (out of 3000) that fix the issue here, isn't trivial at all and I fully understand if you cannot take that effort.
